src/client/Client.cc

```diff
@@ -12137,6 +12137,7 @@ void Client::C_nonblocking_fsync_state::advance()
 
     if (waitfor_safe) {
       clnt->put_request(req);
+      waitfor_safe = false;
```

Comment:
This seems transparently correct since we invoke ```req->get();``` in the previous block when setting ```waitfor_safe = true;```, and no part of C_nonblocking_fsync_state_advancer changes the refcount.

But do we then need to keep more refs around in the following block? What's the ref holding it alive when we ```clnt->add_nonblocking_onfinish_to_context_list(in->waitfor_commit, advancer);```?

Reply:
So, that depends on which event the execution is waiting for. `waitfor_safe` is a part of MetaRequest and calls that unblock those waiters also touch reference counts for MetaRequest. So, a ref is grabbed when blocked for `waitfor_safe.` OTOH, `waitfor_commit` is a part of Inode and that doesn't require a ref to be grabbed on the request.

Reply:
Oh I had totally misunderstood here. This ```req``` is only for the metadata ops list, which we only track if we are ```!syncdataonly```. So once we have gotten a metadata flush that far, we are good to go.

I looked at it a bit more broadly and the way ```C_nonblocking_fsync_flush_finisher``` and ```C_nonblocking_fsync_state_advancer``` interact with each other and case 1 scared me, but I think it actually works.

Reply:
> Looking through the code today and this discussion, the proposed fix looks reasonable to me. Do you have an ETA for reproducer?

Yes -- will get it done today!